### PR TITLE
Fix Nav block editing wrong entity on creation of new Menu

### LINKED ISSUE #36862: Navigation block updates the wrong navigation entity record
### Description

- Delete all menus.
- Site Editor.
- Create Nav block and create a Menu adding 1 item.
- Save.
- Go back to Nav block and Select menu -> Create new.
- Add some items to this new Menu.
- Save.
- Notice how both menus are now marked as dirty.
- Saving causes the original / first menu to loose all its items.

Found by @getdave  in https://github.com/WordPress/gutenberg/pull/36828#pullrequestreview-815880665

### PR DESCRIPTION
**Pull request by getdave**
## Description
In https://github.com/WordPress/gutenberg/issues/36862 we learnt that if you use the `Create new` buttons within the Nav block UI to create a new Menu you can end up causing edits to the wrong `wp_navigation` entity.

How does this happen? Scenario:

- you have an existing Nav block with an existing entity that is being edited
- you click `Create new` drop the `Select menu` dropdown on the block
- at this point the code tries to reset the inner blocks of the block and give you a fresh `wp_navigation` entity for editing.
- to do this it calls `replaceInnerBlocks` but it does this before the entity is set to the freshly created one. Instead it ends up calling replaceBlocks and the _existing_ entity picks up the changes and gets marked as "dirty"
- this process isn't obvious to the user as all the blocks are nicely reset in the UI and everything looks good
- the users starts adding blocks and then hits `Save` in the Site Editor only to find they now have to save both the new and the original Nav Post entities

This PR fixes that by moving the reseting of the inner blocks to an effect which runs if the `ref` to the `wp_navigation` Post is reset. This way we ensure we have completely "unloaded" the current entity before we reset the blocks in the UI. 

⚠️ I worked on this late in my day so it will need some careful testing and reviewing to ensure I'm not introducing errors.

Closes https://github.com/WordPress/gutenberg/issues/36862

## How has this been tested?

* Delete all Navigation Menu posts.
* Go to Site Editor and add a Nav block.
* Add a few items.
* Save the Site Editor - you should see single Nav entity is being saved. Take note of the name of the entity.
* Go back to Nav block and click `Select menu -> Create new`.
* When the block is ready start adding new items.
* Save the Site Editor - you should see only a _single_ Nav entity is being saved. It should be different from the one you originally created.

## Types of changes
Bug fix (non-breaking change which fixes an issue) 

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] I've tested my changes with keyboard and screen readers. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

File: packages/e2e-tests/specs/experiments/blocks/navigation.test.js [hunk 1]
```diff
@@ -784,4 +784,92 @@ describe.skip( 'Navigation', () => {
 
 		expect( isScriptLoaded ).toBe( true );
 	} );
+
+	describe( 'Creating and restarting', () => {
```
Comment (talldan) [inline packages/e2e-tests/specs/experiments/blocks/navigation.test.js:788]: Navigation e2e tests are currently all skipped, so this won't run.

I did start rewriting them completely, but have been too busy to finish up (plus everything is changed too rapidly to keep up.)

Comment (getdave) [reply packages/e2e-tests/specs/experiments/blocks/navigation.test.js:788]: Ah yes we forgot that. I pulled in from Adam's PR. We can leave it in for now especially since you unfocused it.

**Comment by getdave** [comment]
Failing e2e test appears to be around dirty entities which worries me. I'll run locally to check...

**Comment by getdave** [comment]
> Failing e2e test appears to be around dirty entities which worries me. I'll run locally to check...

I think we've introduced a bug here with dirty entity states as the test is failing because the `header` template part has changes when loading the Site Editor.

**Comment by adamziel** [comment]
I added an additional check for `&& innerBlocks.length > 0` to fix that test. Let's see what it does to all the other ones 🤞  @getdave

File: packages/block-library/src/navigation/edit/index.js [hunk 2]
```diff
@@ -279,14 +279,24 @@ function Navigation( {
...
 
+	// If the ref no longer exists the reset the inner blocks
+	// to provide a clean slate.
+	useEffect( () => {
+		if ( ref === undefined && innerBlocks.length > 0 ) {
+			replaceInnerBlocks( clientId, [] );
+		}
+		// innerBlocks are intentionally not listed as deps. This function is only concerned
+		// with the snapshot from the time when ref became undefined.
+	}, [ clientId, ref ] );
```
Comment (getdave) [inline packages/block-library/src/navigation/edit/index.js]:
```suggestion
	}, [ clientId, ref, innerBlocks ] );
```

Do we now need to add `innerBlocks` to the deps?

**Comment by getdave** [comment]
@adamziel this now looks good. e2e passes. Needs a 👍 pending one final query in https://github.com/WordPress/gutenberg/pull/36880#discussion_r757457035

File: packages/block-library/src/navigation/edit/index.js (hunk 2 above)
Comment (adamziel) [reply packages/block-library/src/navigation/edit/index.js]: No, as the comment above states :-)

**Review by adamziel — APPROVED** [review]

File: packages/block-library/src/navigation/edit/index.js [hunk 3]
```diff
@@ -279,14 +279,24 @@ function Navigation( {
...
 
+	// If the ref no longer exists the reset the inner blocks
+	// to provide a clean slate.
+	useEffect( () => {
+		if ( ref === undefined && innerBlocks.length > 0 ) {
+			replaceInnerBlocks( clientId, [] );
+		}
+		// innerBlocks are intentionally not listed as deps. This function is only concerned
+		// with the snapshot from the time when ref became undefined.
+	}, [ clientId, ref, innerBlocks ] );
```
Comment (talldan) [inline packages/block-library/src/navigation/edit/index.js:290]: I think this has caused a potential bug around pattern insertion in the nav block, the nav block no longer supports UnsavedInnerBlocks. Every time the block has Inner Blocks but no ref, the inner blocks are deleted straight away.

It is a little hard to follow, so I fully appreciate the difficulty level in understanding why this would be a problem.

Also just wanted to flag the comment - 'innerBlocks are intentionally not listed as deps.' - they are listed as deps?

Comment (talldan) [reply packages/block-library/src/navigation/edit/index.js:290]: I've created a bug report for this - https://github.com/WordPress/gutenberg/issues/37354

File: packages/block-library/src/navigation/edit/index.js (hunk 2 above)
Comment (talldan) [reply packages/block-library/src/navigation/edit/index.js]: Oops, looks like it was added anyway in 793ff2a

Comment (getdave) [reply packages/block-library/src/navigation/edit/index.js]: I don't even know what happened here 😕

File: packages/block-library/src/navigation/edit/index.js (hunk 3 above)
Comment (getdave) [reply packages/block-library/src/navigation/edit/index.js:290]: Thanks Dan and apologies for any problems.

I guess this is why we need stable tests.
